### PR TITLE
Enable Globalization Invariant mode using Environment Variable

### PR DESCRIPTION
**Pull request by tarekgh**

### CODE REVIEW

**Comment by tarekgh** [comment]
https://github.com/dotnet/coreclr/issues/14927

**Comment by tarekgh** [comment]
@jkotas could you please have a look? This change will enable supporting the environment variable. I'll look at enabling the invariant mode by default in Alpine build next.

**Comment by tarekgh** [comment]
CC @Petermarcu @MichaelSimons

File: src/mscorlib/src/System/Globalization/GlobalizationMode.cs [hunk 1]
```diff
@@ -7,6 +7,27 @@ namespace System.Globalization
     internal sealed partial class GlobalizationMode
     {
         private const string c_InvariantModeConfigSwitch = "System.Globalization.Invariant";
+        // Linux doesn't support environment variable names including dots
+        private const string c_InvariantModeEnvironmentVariable = "System_Globalization_Invariant";
```
Comment (jkotas) [inline src/mscorlib/src/System/Globalization/GlobalizationMode.cs]: We need to get a standardized convention for the environment variables. We have `COMPlus_...`, `DOTNET_...`, ...  it is all over the place.

Is there any other prior art that is using `System_...` prefix?

Comment (stephentoub) [reply src/mscorlib/src/System/Globalization/GlobalizationMode.cs]: I agree we need to standardize.  And DOTNET_ seems like the best choice, as from a consumer's perspective their really shouldn't be a difference between coreclr, corefx, etc.  Most of the existing environment variables are COMPlus_... would there be any concerns changing them?

Comment (tarekgh) [reply src/mscorlib/src/System/Globalization/GlobalizationMode.cs]: DOTNET_ prefix seam reasonable to me. if all agree, I can change the name here. we can log an issue too to track renaming the other introduced environment variable.

P.S. I don't like COMPLUS_ nor CORCLR _ prefixes as most of people don't know what is complus and also CORECLR can give the impression this is specific to coreclr repo (and never apply for corert for instance)

**Comment by tarekgh** [comment]
test Tizen armel Cross Checked Innerloop Build and Test please

File: src/mscorlib/src/System/Globalization/GlobalizationMode.cs [hunk 2]
```diff
@@ -7,6 +7,27 @@ namespace System.Globalization
     internal sealed partial class GlobalizationMode
     {
         private const string c_InvariantModeConfigSwitch = "System.Globalization.Invariant";
+        // Linux doesn't support environment variable names including dots
+        private const string c_InvariantModeEnvironmentVariable = "DOTNET_System_Globalization_Invariant";
```
Comment (jkotas) [inline src/mscorlib/src/System/Globalization/GlobalizationMode.cs]: The existing `DOTNET_` variables are all upper case: `DOTNET_RUNTIME_SERVICING`, `DOTNET_PACKAGES`, `DOTNET_SKIP_FIRST_TIME_EXPERIENCE`, ... . Do we want to be consistent here? (I am not sure ... just asking.)

File: src/mscorlib/src/System/Globalization/GlobalizationMode.cs [hunk 3]
```diff
@@ -7,6 +7,27 @@ namespace System.Globalization
...
         internal static bool Invariant { get; } = GetGlobalizationInvariantMode();
+
+        // GetInvariantSwitchValue calls CLRConfig first to detect if the switch is defined in the config file.
+        // if the switch is defined we just use the value of this switch. otherwise, we'll try to get the switch
+        // value from the environment variable if it is defined.
+        internal static bool GetInvariantSwitchValue()
+        {
+            bool exist;
+            bool ret = CLRConfig.GetBoolValue(c_InvariantModeConfigSwitch, out exist);
```
Comment (jkotas) [inline src/mscorlib/src/System/Globalization/GlobalizationMode.cs]: Nit: I think that constants for strings that are used exactly once are negative value.

File: src/mscorlib/src/System/Globalization/GlobalizationMode.cs [hunk 4]
```diff
@@ -7,6 +7,27 @@ namespace System.Globalization
...
+            if (!exist)
+            {
+                string switchValue = Environment.GetEnvironmentVariable(c_InvariantModeEnvironmentVariable);
+                if (switchValue != null)
+                {
+                    ret = switchValue.Equals("true", StringComparison.OrdinalIgnoreCase);
```
Comment (jkotas) [inline src/mscorlib/src/System/Globalization/GlobalizationMode.cs]: It is more common to use 0/1 for these. (We can recognize true/false as well, but 0/1 has to work.)

**Comment by jkotas** [comment]
LGTM otherwise

**Comment by tarekgh** [comment]
@dotnet-bot test this

**Comment by janvorli** [comment]
@tarekgh it would be great to add description of this variable to our documentation where we describe how to turn on the invariant globalization.

**Comment by tarekgh** [comment]
@janvorli I already asked @richlander in the following comment to update the doc he originally published. 

https://github.com/dotnet/coreclr/issues/14927#issuecomment-344664479

I can help update the document too if @richlander want me to do it.

**Pull request merged**